KEP 4770: EndpointSlice Controller Flexibility (#4771)
kubernetes/enhancements: keps/sig-network/4770-endpointslice-controller-flexibility/README.md

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

This proposal adds a new well-known label `service.kubernetes.io/endpoint-controller-name` to Kubernetes Services. This label disables the default Kubernetes EndpointSlice, EndpointSlice Mirroring and Endpoints controllers for the services where this label is applied and delegates the control of EndpointSlices to a custom EndpointSlice controller.

## Motivation

As of now, a service can be delegated to a custom Service-Proxy/Gateway if the label `service.kubernetes.io/service-proxy-name` is set. Introduced in [KEP-2447](https://github.com/kubernetes/enhancements/issues/2447), this allows custom Service-Proxies/Gateways to implement services in different ways to address different purposes / use-cases. However, the EndpointSlices attached to this service will still be reconciled in the same way as any other service. Addressing more purposes / use-cases, for example, different pod IP addresses, is therefore not natively possible.

> Delegating EndpointSlice control would allow custom controllers to define their own criteria for pod availability, selecting different pod IPs than the pod.status.PodIPs and more. As a reference implementation, and since [KEP-3685](https://github.com/kubernetes/enhancements/issues/3685), the reconciler logic used by Kubernetes can be reused by custom EndpointSlice controllers.

Reviewer: You should explain what KEP-3685 is rather than just referring to it by number, but I think this is mostly talking about the withdrawn plan to extend the existing reconciler anyway, so probably it should have been deleted?

Author: Thank you, I added the short description of the KEP. No, I think this is an additional argument as an implementation is required when using the label, so the EndpointSlice Reconciler in staging could be used as reference.

### Goals

* Provide the ability to disable the Kubernetes EndpointSlice, EndpointSlice Mirroring and Endpoints controllers for particular services.

### Non-Goals

* Change / Replace / Deprecate the existing behavior of the Kubernetes EndpointSlice, EndpointSlice Mirroring and Endpoints controllers.
* Introduce additional supported types of the EndpointSlice controllers/Reconciler as part of Kubernetes.
* Modify the Service / EndpointSlice / Endpoints Specs.

> * Add new features to the EndpointSlice Mirroring Controller.

Reviewer: I think it would be useful as a clarification to add "Changing the behavior of kube-proxy" as a Non-Goal: Kube-proxy doesn't need to care who creates the EndpointSlices for a Service or what logic they use to do so. It just knows that, unless the service has a `service-proxy-name` label indicating otherwise, it should just create rules pointing to its endpoints. Right?

And likewise cloud load balancers?

Author: Thank you, I added it. Yes, this applies to any service (described in the next section: "When set on a service, no matter the service specs...")

## Proposal

#### Well-Known Label

`service.kubernetes.io/endpoint-controller-name` will be added as a well-known label applying on the Service object. When set on a service, no matter the service specs, the Endpoint, EndpointSlice, and EndpointSlice Mirroring controllers for that service will be disabled, thus Endpoints and EndpointSlices for this service will not be created by the Kubernetes Controller Manager. If the label is not set, the Endpoint, EndpointSlice, and EndpointSlice Mirroring controllers will be enabled for that service and the Endpoints and EndpointSlices will be handled as of today.

The Kubernetes Controller Manager will implement this label both at object creation and on dynamic addition/removal/updates of this label.

### User Stories (Optional)

> #### Story 1

Reviewer: I think it's considered good style to give the story a "title". ("Story 1: Services on Multus Networks"?)

Author: I fixed it, thank you.

As a Cloud Native Network Function (CNF) vendor, some of my services are handled by custom Service-Proxies/Gateways over secondary networks provided by, for example, [Multus](https://github.com/k8snetworkplumbingwg/multus-cni). IPs configured in the service and registered by the EndpointSlice controller must be only the secondary IPs provided by the secondary network provider (e.g. [Multus](https://github.com/k8snetworkplumbingwg/multus-cni)).

> Therefore, it must be possible to disable the default Kubernetes Endpoints and EndpointSlice Controller for certain services and use a specialized EndpointSlice reconciler implementation to create a controller for secondary network providers.

Reviewer: It's not clear where you're using "service" to mean `v1.Service` and where you're using it in a generic sense...

How is the Service configured? Is it `type: ClusterIP`? Does it have the `service-proxy-name` label set?

Who is calling the Service and from where? (Pods that share the same secondary network? Pods that _don't_ share the same secondary network? Nodes? Cluster-external hosts?) What is the expectation for if the "wrong" people try to connect to it?

(You have both "for example, [Multus]" and "e.g. [Multus]" there btw. Second one can probably go away.)

Author: I always meant the Kubernetes service, I added it for clarity, thank you.
I also added the mention to the service-proxy label. But I think the rest of the configuration of the service doesn't really matter, it could be both internal service traffic or traffic originating from outside of the cluster. I am not sure we should try to fully describe secondary networks and services here.

Reviewer: > But I think the rest of the configuration of the service doesn't really matter

Ideally, user stories should help the reader understand why you made some decisions and not others. But they can only do that if the user stories are sufficiently clear and evocative.

(And you _especially_ don't want your user stories to be making the reviewers think about things like "but how will `type: LoadBalancer` services work if the load balancer doesn't have access to the secondary network?")

### Notes/Constraints/Caveats (Optional)

> N/A

Reviewer: (if it's "N/A" you can just remove the section, since it's "(Optional)")

Author: I will remove them later, if anyone has any concern that some are in fact applicable.

### Risks and Mitigations

The existing behavior will be kept by default, and the Kubernetes EndpointSlice Controller will not manage the Services with the label. This ensures services without the label to continue to be managed as usual.

This will have no effect on other EndpointSlice controller implementations since they will not be influenced by the presence of this label.

## Design Details

### Well-Known Label

> The kube-controller-manager will pass to the Endpoints, EndpointSlice and EndpointSlice Mirroring Controllers an informer selecting services that are not labeled with `service.kubernetes.io/endpoint-controller-name`. Thus, if the label is added to an existing service (by updating the service), the service with the label will be considered as a deleted service for the controllers, and the Endpoints and EndpointSlices will be deleted. If a Service is created with the label, the controllers will not be informed about it, so the Endpoints and EndpointSlices will not be created. If the label is removed from an existing service (by updating the service), the service with the label will be considered as a newly created service for the controllers, and the Endpoints and EndpointSlices will be created.

Reviewer 2: @robscott @danwinship @thockin The mutability here is a friction point, as this is "the same" as removing the selector on a Service , something we didn't reach a consensus yet and this is defining a behavior that is inconsistent with current state https://github.com/kubernetes/kubernetes/issues/118376.

I'm leaning more to avoid to deal with Updates on Services, specially because I can not understand why one Service with one selector managed by the cluster endpoint slice controller will switch to an external one slice controller, it seems that should be better handled by deleting and recreating the Service.

This will be straight forward to implement, just validating on the Service API that this special label can not be added or mutated after the Service has been created

Reviewer: If the annotation is mutable (which it has to be) then we have to deal with changes to it. Even if kcm tries to ignore changes, what happens if you restart kcm after changing the annotation?

I feel like 118376 was mostly waiting for Rob to bless a solution, and then we all stopped thinking about it.

Author: I am not sure if I miss something. To me, 118376 is different issue than the one describe in this KEP.

Here is how I see the implementation for this KEP: The service informer passed to the EndpointSlice Controller will have a selector to not include the services that are labeled with `service.kubernetes.io/endpoint-controller-name`. This means the controller will consider adding the label as if the service has been removed (which should already be covered as of now: when you remove a service, the EndpointSlices for it are removed).

If KCM is restarting, then the EndpointSlice controller will see the previous EndpointSlice(s) and enqueue the service associated (service name is in the EndpointSlice annotation). The service will be considered as non-existing (as the informer is not selecting labeled service), so the EndpointSlice(s) will be deleted.

Here is the implementation: https://github.com/kubernetes/kubernetes/compare/master...LionelJouin:kubernetes:custom-endpointslice-controller

Reviewer: The similarity between this and 118376 is that in both cases someone (potentially) has a Service with endpoints that have been generated in a particular way, and then later they want to change the Service so that the endpoints will be generated in a different way. Antonio was suggesting that if we couldn't figure out a solution for 118376, then we won't be able to come up with a solution for this either (and so we should forbid the case where you change things after the endpoints have already been generated), and I was disputing that we couldn't figure out a solution for 118376.

Author: I saw this comment: https://github.com/kubernetes/kubernetes/issues/103576#issuecomment-876844762. The behavior that I see with this KEP is different. To me, the first interpretation seems more logical for this KEP, so if we add the label, controllers are disabled for the service, and objects (Endpoints and EndpointSlices) previously created by the KCM are deleted.

Reviewer 2: > If the annotation is mutable (which it has to be) 

I think you mean label instead of annotation ...  why it has to be mutable?

> if we add the label, controllers are disabled for the service, and objects (Endpoints and EndpointSlices) previously created by the KCM are deleted

> Here is the implementation: https://github.com/kubernetes/kubernetes/compare/master...LionelJouin:kubernetes:custom-endpointslice-controller

I see, so this is using this property of the informers so the controller "sees" a Delete event when the label is added, with 118376 the problem is that it "sees" an Update event ... I can not think of any edge cases from the top of my head, I think it may work ...

@robscott WDYT? maybe should we modify the controllers to process the selector removal as a delete event too?

    * A service is created without the label, the service has endpoints and endpointslices. Then service is updated with the label and the service has no longer any endpoints nor endpointslices.

##### e2e tests

TDB

### Graduation Criteria

#### Alpha

- Feature implemented behind feature gates (`EndpointControllerNameWellKnownLabel`). Feature Gates are disabled by default.
- Documentation provided.
- Initial unit, integration and e2e tests completed and enabled.

#### Beta

- Feature Gates are enabled by default.
- No major outstanding bugs.
- Feedback collected from the community (developers and users) with adjustment provided, implemented and tested.

#### GA

- 2 examples of real-world usage.
- Allowing time for feedback from developers and users.

### Upgrade / Downgrade Strategy

N/A

### Version Skew Strategy

N/A

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: `EndpointControllerNameWellKnownLabel`
  - Components depending on the feature gate: kube-controller-manager
- [x] Other

>   - Describe the mechanism: setting the label `service.kubernetes.io/endpoint-controller-name` on a service disables the Kubernetes endpointslice controller for that particular service. So no EndpointSlices/Enpoints for this service from the kube-controller-manager will be existing.

Reviewer: That's not "enabling/disabling" the feature, that's _using_ the feature. The "feature gate" checkbox above is correct and is all you need here.

Author: I fixed it, thank you.

  - Will enabling / disabling the feature require downtime of the control
    plane? No
  - Will enabling / disabling the feature require downtime or reprovisioning
    of a node? No

###### Does enabling the feature change any default behavior?

> When the feature-gate `EndpointControllerNameWellKnownLabel` is enabled, the label `service.kubernetes.io/endpoint-controller-name` will work as described in this KEP.

Reviewer: I think the intended answer here is "no, because existing services will not have the endpoint-controller-name label, so the endpoint controllers will continue to generate endpoints for all services"

Author: I fixed it, thank you.

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

> N/A

Reviewer: Some of these aren't N/A.

e.g., if a cluster is making use of this feature, and you roll back to a version of kcm that doesn't know about the label, then that kcm will start overwriting your custom endpoints. So that's something we need to warn users about.

Author: I added it for disabling and re-enabling the feature, I will have to check the testing.

Author: I added a test in the integration tests for when the feature gate is enabled and disabled.

Delegating EndpointSlice control would allow custom controllers to define their own criteria for pod availability, selecting different pod IPs than the pod.status.PodIPs and more. As a reference implementation, and since the EndpointSlice Reconciler has been moved into Staging in [KEP-3685](https://github.com/kubernetes/enhancements/issues/3685), the reconciler logic used by Kubernetes can be reused by custom EndpointSlice controllers.

### Goals

* Provide the ability to disable the Kubernetes EndpointSlice, EndpointSlice Mirroring and Endpoints controllers for particular services.

### Non-Goals

* Change / Replace / Deprecate the existing behavior of the Kubernetes EndpointSlice, EndpointSlice Mirroring and Endpoints controllers.
* Introduce additional supported types of the EndpointSlice controllers/Reconciler as part of Kubernetes.
* Modify the Service / EndpointSlice / Endpoints Specs.
* Changing the behavior of kube-proxy.

## Proposal

#### Well-Known Label

> `service.kubernetes.io/endpoint-controller-name` will be added as a well-known label applying on the Service object. When set on a service, no matter the service specs, the Endpoint, EndpointSlice, and EndpointSlice Mirroring controllers for that service will be disabled, thus Endpoints and EndpointSlices for this service will not be created by the Kubernetes Controller Manager. If the label is not set, the Endpoint, EndpointSlice, and EndpointSlice Mirroring controllers will be enabled for that service and the Endpoints and EndpointSlices will be handled as of today.

Reviewer 2: more on nameing , can the label name be misleading ?

Reviewer: It seems to fit very well with the existing `service.kubernetes.io/service-controller-name`.

Unless you mean specifically the word "endpoint" as opposed to "endpointslice", but I think at this point, endpoint slices _are_ "endpoints", and we don't need to still always be explicit about that in naming.

Reviewer 2: yeah, it was about "endpoint", I think your argument is very convincing

    * With the feature gate enable, a service is created with the label and the service has then no Endpoints neither EndpointSlices. Then service is updated removing the label and the service has now Endpoints and EndpointSlices.
    * With the feature gate enable, a service is created without the label, the service has Endpoints and EndpointSlices. Then service is updated with the label and the service has no longer any Endpoints nor EndpointSlices.
    * With the feature gate disabled, a service is created with the label and the service has EndpointSlices as it would have had without the label.

##### e2e tests

- Usage of `service.kubernetes.io/endpoint-controller-name` on services
    * A service is created with the label, an EndpointSlice is manually created with an endpoint (simulating an external controller). The test verifies the service is reachable. 

### Graduation Criteria

#### Alpha

- Feature implemented behind feature gates (`EndpointControllerNameWellKnownLabel`). Feature Gates are disabled by default.
- Documentation provided.
- Initial unit, integration and e2e tests completed and enabled.

#### Beta

> 

Reviewer 2: There has to be some e2e conformance of these out of tree controllers, since is an external controller adding an endpointslice, how do we validate these out of tree controllers are generating the endpointslices accordingly ? what if these controllers does not respect the selectors and generete the endpointslices anyway?

@robscott I really like to hear your thoughts on this

Reviewer 3: That's an interesting thought. I'd assumed that the purpose of this KEP was to build controllers that would intentionally have different behavior than the default in-tree EndpointSlice controller. In some cases that might be to select a different IP for a Pod, but in other cases that might be to select an entirely different type of resource, for example a custom backend type. 

This does certainly make it more likely that unexpected variations of EndpointSlices will exist, which is certainly something to watch out for. Since the outset of the EndpointSlice API _anyone_ has been able to create EndpointSlices, and sometimes unexpected variations have led to bugs in controllers consuming those EndpointSlices, but I don't think that's necessarily new with this KEP, just potentially a more common set of edge cases that we'll need to be sure that controller authors are aware of. 

For example, `nodeName` is optional in EndpointSlices, but some controllers likely assume it exists because the EndpointSlice controllers in k/k _generally_ populate it. The same is largely true for zone and hostname (see [podToEndpoint](https://github.com/kubernetes/kubernetes/blob/7df2b05e772c4190f53038cf529ef00769818899/staging/src/k8s.io/endpointslice/utils.go#L58-L69) for a reference).

I think this risk is unavoidable with this KEP, so I'd like to see it called out + a corresponding commitment to improved documentation around this as a requirement for graduating past alpha. Unfortunately I can't think of a way to write conformance tests that would help us here though.

/cc @swetharepakula @gauravkghildiyal

Reviewer 4: I wonder if we want to test that EndpointSlices created before a Service still work how we want them to (kube-proxy behavior for example), even after that Service is eventually added?

### Well-known label: Use Annotation as Selector 

Services without selectors will not get any EndpointSlice objects. Therefore, selecting pods can be done in different ways, for example, via annotation. An annotation will be used in the service to select which pods will be used as backend for this service. For example, [nokia/danm](https://github.com/nokia/danm) uses `danm.k8s.io/selector` (e.g. [DANM service declaration](https://github.com/nokia/danm/blob/v4.3.0/example/svcwatcher_demo/services/internal_lb_svc.yaml#L7)), and [projectcalico/vpp-dataplane](https://github.com/projectcalico/vpp-dataplane) uses `extensions.projectcalico.org/selector` (e.g. [Calico-VPP Multinet services](https://github.com/projectcalico/vpp-dataplane/blob/v3.25.1/docs/multinet.md#multinet-services)). To simplify the user experience, a mutating webhook could read the selector, add them to the annotation and clear them from the specs when the type of service is detected.

The custom EndpointSlice Controller will then read the annotation to select the pods targeted by the service.

```yaml
apiVersion: v1
kind: Service
metadata:
  name: service
  annotations:
    selector: "app=a"
spec: {}
```

> This alternative potentially leads to confusion among users and inconsistency in how services are managed as each implementation is using its own annotation (see the nokia/danm and projectcalico/vpp-dataplane examples), leading to a fragmented approach.

Reviewer 5: My initial reaction to this KEP is that we should not do it.  This alternative is better for a bunch of reasons, but mostly because it already exists.  

We have a way of expressing "don't generate endpoints for me" - leave the selector blank.  Fortunately for you, it means exactly what you want.

The argument of inconsistency across controllers doesn't hold a lot of water for me.  If you specify a pod selector, you're going to get pods as endpoints.  If you don't specify it, you get something else.  IMO, being able to specify that something else in your own schema is an advantage, not "fragmentation".  Suppose you need an expression that is more flexible than the legacy selector-map (e.g. modern selectors can use "In" and "NotIn" and other operators). 

This is not an accidental side-effect that I am trying to shoe-horn you into, it's literally why this mechanism was designed.

Author: The empty selector feature exists to cover use-cases that are not covered by the existing selector, i.e. selecting backends that are not selectable via labels. For example ([from the Service documentation](https://kubernetes.io/docs/concepts/services-networking/service/#services-without-selectors)), selecting backends in a different namespace/cluster, selecting backends that are running outside of a Kubernetes, or as you mentioned, selecting backends with a more flexible expression covered with modern selectors. In these cases I agree the empty selector must be used and the users must define their own way to select backends, via, for example, annotations.

I realized that the goals and non-goals are not clear enough, so I fixed them in the latest commit. The goal is to disable the Kubernetes EndpointSlice controller and allow the user to re-use the selector field from the Service Spec. Custom EndpointSlice Controller should be able to replicate the usage of the selector, by selecting workloads by labels (labels specified in the selector field) but the difference with the Kubernetes EndpointSlice Controller is the compilation of these EndpointSlices, in this KEP the main use case is to select different IPs than the Pod.Status.PodIPs.

A well-known label would also provide an explicit and standard way to inform which component is managing the EndpointSlices for a particular service.

Reviewer 5: I am still struggling with this one.  I spoke with Rob a little, since he seemed "pro".  The argument that carried the most weight for me was that there are (or might be?) cases where the alternative controller is doing exactly what the "real" controller does but extracting some different information (e.g. the multi-network case).  In that case, you have to specify the network(s) somewhere right?  Another example, albeit hypothetical, was something doing different topology hinting.

I'm not sure how I feel about this, still.  It seems "simple" to implement, but I am worried about the tail of side-effects.

As a thought experiment - what if you could specify a function, e.g. CEL, which took a pod as argument and returned a list of IPs to use -- would that eliminate the need for this?

Author: Yes, that's correct, multi-network is my primary use case, I have personally no plan, at least for now, to do something different such as topology hinting. But, multi-network also means more than just picking the IP for the pods, it also means the network must be attached and up for the selected pods. 

Yes, I need to specify which network(s) must be used for the selected pods. Having this label (`service.kubernetes.io/endpoint-controller-name`) will give me freedom to choose the way to specify those networks. For my case, I am currently thinking of using Services together with Gateway API (I made a PoC about it: https://github.com/LionelJouin/l-3-4-gateway-api-poc). Gateway in Gateway API allows me to specify new Service-Proxies (Secondary Network Service (load-balancing) acts directly in the Gateway instance) and attach networks via the Gateway.Spec.Infrastructure field, this way I can also specify which networks must be selected. Here is an example on how the Service/Gateway could look like: https://github.com/LionelJouin/l-3-4-gateway-api-poc/blob/main/examples/kpng-gateway-api.yaml#L59-L62.

If I understand correctly, CEL would select a specific field based on what the user would specify via a new field in the Service. I think this would be more complex to maintain, more difficult to use, and less flexible. As of today, the most used solution is [Multus](https://github.com/k8snetworkplumbingwg/multus-cni), the secondary networks are attached and reported via pod annotation in a json format, I guess it would be quite challenging to match the IPs for a particular network. This could have worked with the [Multi-Network KEP](https://github.com/kubernetes/enhancements/pull/3700) changes since the Secondary Network IPs would have been reported in Pod.Status. Now the discussion for multi-network is about DRA and [ResourceClaim.Status to report the Secondary Network IPs](https://github.com/kubernetes/enhancements/issues/4817), I am not sure how CEL could fit here.

For reference, at the beginning, another goal for this KEP was to improve the Kubernetes EndpointSlice Reconciler to make it generic enough so it can be re-used for Secondary Network IPs (No matter if Multus, KEP#3698 or DRA is used). This way, the custom EndpointSlice Controllers for secondary networks would be quite easy to implement. I kept it in the commit history ([removed in this commit](https://github.com/kubernetes/enhancements/pull/4771/commits/51fd477a42d4a99252eb907c32fa1e2d0bda63da)) and I am still planning to propose again these changes somehow in a better way in the future.

Reviewer 2: > For reference, at the beginning, another goal for this KEP was to improve the Kubernetes EndpointSlice Reconciler to make it generic enough so it can be re-used for Secondary Network IPs (No matter if Multus, KEP#3698 or DRA is used). This way, the custom EndpointSlice Controllers for secondary networks would be quite easy to implement. I kept it in the commit history ([removed in this commit](https://github.com/kubernetes/enhancements/pull/4771/commits/51fd477a42d4a99252eb907c32fa1e2d0bda63da)) and I am still planning to propose again these changes somehow in a better way in the future.

This is a big no for me, if someone wants to create EndpointSlices controllers out of tree is ok to share some common types and logic but adding more load to the core maintainers just to remove it the custom endpointslice controllers maintainers is simply unfair.

Author: > This is a big no for me, if someone wants to create EndpointSlices controllers out of tree is ok to share some common types and logic but adding more load to the core maintainers just to remove from the custom endpointslice controllers maintainers is simply unfair.

I agree with this, the idea with this part (generic EndpointSlice Reconciler) of the KEP was to solve what is stated in the [readme of EndpointSlice Reconciler](https://github.com/kubernetes/endpointslice): `This EndpointSlice reconciler library is not sufficiently generic to be used by the EndpointSlice Mirroring controller...`. But also keeping what is written for compatibility: `There are NO compatibility guarantees for this repository, yet. It is in direct support of Kubernetes, so branches will track Kubernetes and be compatible with that repo.`.

- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

This proposal adds a new well-known label `service.kubernetes.io/endpoint-controller-name` to Kubernetes Services. This label disables the default Kubernetes EndpointSlice, EndpointSlice Mirroring and Endpoints controllers for the services where this label is applied and delegates the control of EndpointSlices to a custom EndpointSlice controller.

## Motivation

> As of now, a service can be delegated to a custom Service-Proxy/Gateway if the label `service.kubernetes.io/service-proxy-name` is set. Introduced in [KEP-2447](https://github.com/kubernetes/enhancements/issues/2447), this allows custom Service-Proxies/Gateways to implement services in different ways to address different purposes / use-cases. However, the EndpointSlices attached to this service will still be reconciled in the same way as any other service. Addressing more purposes / use-cases, for example, different pod IP addresses, is therefore not natively possible.

Reviewer 4: Is this true? What if you define a Service without a selector and then do your own reconciliation?

Author: Thank you for your comment, I clarified that it is not natively possible **while using the Service Selector field** (The KEP is about this, the fact to use the service as close as possible to how a user would define it for the Kubernetes default network).
About this topic, there are details in this conversation: https://github.com/kubernetes/enhancements/pull/4771#discussion_r1752766918

spec:
  selector:
    app: a
    dummy-selector: "true"
```

This alternative fails to prevent the placeholder (empty) EndpointSlice(s) to be created by Kube-Controller-Manager. This also potentially causes confusion among users as every implementation could use a different dummy-selector key. Additionally, a miss-configuration with the missing dummy label will lead to unintended EndpointSlices being created with Pod.Status.PodIPs.

### Well-known label: Disable the Kube-Controller-Manager Controllers

The list of controllers to enable in the Kube-Controller-Manager can be set using the `--controllers` flag ([kube-controller-manager documentation](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/)). The EndpointSlice can then be disabled in the Kube-Controller-Manager and implemented as an external one that will support the label feature described in this KEP.

This alternative requires significant changes to the cluster management as the cluster level configuration must be modified and a new EndpointSlice controller for the primary network must be developed and deployed to replace the disabled one in the Kube-Controller-Manager.

> 

Reviewer 4: We should - I pretty much think **must** - document the "don't specify a selector, then do your own thing" option. It _is_ an alternative and we shouldn't pretend otherwise.

Author: This is covered by the first alternative `Well-known label: Empty Selector field and Use Annotation as Selector`, I renamed the alternative so it is more clear.

Reviewer 4: BTW you can't select on annotations.

Author: you can create your own selector in the annotation, here is an example.
```yaml
apiVersion: v1
kind: Service
metadata:
  name: service
  annotations:
    selector: "app=a"
spec: {}
```

Reviewer 4: Noted

In the Endpoints, EndpointSlice and EndpointSlice Mirroring Controllers, the behavior to create Endpoints/EndpointSlices on service creation and the behavior to delete the Endpoints/EndpointSlices on service deletion is already in place. Only the service informer passed to these controllers must be tweaked for the proposed well-known label (`service.kubernetes.io/endpoint-controller-name`) to work properly.

> This new well-known label will not have any impact on the other well-known label `endpointslice.kubernetes.io/managed-by`. Both labels can coexist and be used together, ``service.kubernetes.io/endpoint-controller-name` on the Service object while `endpointslice.kubernetes.io/managed-by` on the EndpointSlice object. The label `endpointslice.kubernetes.io/managed-by` is used to indicate the controller or entity that manages the EndpointSlice. In contrast, this new label `service.kubernetes.io/endpoint-controller-name`, is used to disable the default Kubernetes EndpointSlice, EndpointSlice Mirroring and Endpoints controllers for a Service and indicates the controller or entity that manages the EndpointSlice for that particular Service.

Reviewer 4: ```suggestion
This new well-known label will not have any impact on the other well-known label `endpointslice.kubernetes.io/managed-by`. Both labels can coexist and be used together, `service.kubernetes.io/endpoint-controller-name` on the Service object while `endpointslice.kubernetes.io/managed-by` on the EndpointSlice object. The label `endpointslice.kubernetes.io/managed-by` is used to indicate the controller or entity that manages the EndpointSlice(s) for that Service. In contrast, this new label `service.kubernetes.io/endpoint-controller-name`, is used to disable the default Kubernetes EndpointSlice, EndpointSlice Mirroring and Endpoints controllers for a Service and indicates the controller or entity that manages the EndpointSlices for that particular Service.
```

Author: The label `endpointslice.kubernetes.io/managed-by` is used to indicate the controller or entity that manages an EndpointSlice. Adding "for that Service" is more confusing to me as the label `endpointslice.kubernetes.io/managed-by` is applied on an EndpointSlice object, not a Service object.

Reviewer 4: OK, try this:
```markdown
This new well-known label will not have any impact on the other well-known label `endpointslice.kubernetes.io/managed-by`. Both labels can coexist and be used together, `service.kubernetes.io/endpoint-controller-name` on the Service object while `endpointslice.kubernetes.io/managed-by` on the EndpointSlice object. The label `endpointslice.kubernetes.io/managed-by` is used to indicate the controller or entity that manages the EndpointSlice(s) - for, example, the endpoint slice controller managing endpoints on behalf of a Service.
In contrast, this new label `service.kubernetes.io/endpoint-controller-name`, is used to disable the default Kubernetes endpoint slice, Endpoints-to-EndpointSlice mirroring and endpoints controllers for a Service; the value of the new label indicates the controller or entity that manages the EndpointSlices for that particular Service.
```

How does that sound?

### Test Plan

[x] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

##### Unit tests

> TDB

Reviewer 4: ```suggestion
TBD
```

No

### Scalability

###### Will enabling / using this feature result in any new API calls?

No

###### Will enabling / using this feature result in introducing new API types?

No

###### Will enabling / using this feature result in any new calls to the cloud provider?

No

###### Will enabling / using this feature result in increasing size or count of the existing API objects?

No

###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?

No

###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?

No

###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?

No

### Troubleshooting

###### How does this feature react if the API server and/or etcd is unavailable?

N/A

###### What are other known failure modes?

N/A

###### What steps should be taken if SLOs are not being met to determine the problem?

N/A

## Implementation History

- Initial proposal: 2024-07-19

## Drawbacks

TBD

## Alternatives

> ### Well-known label: Empty Selector field and Use Annotation as Selector 

Reviewer 4: OK, resolved then.

> Services without selectors will not get any EndpointSlice objects. Therefore, selecting pods can be done in different ways, for example, via annotation. An annotation will be used in the service to select which pods will be used as backend for this service. For example, [nokia/danm](https://github.com/nokia/danm) uses `danm.k8s.io/selector` (e.g. [DANM service declaration](https://github.com/nokia/danm/blob/v4.3.0/example/svcwatcher_demo/services/internal_lb_svc.yaml#L7)), and [projectcalico/vpp-dataplane](https://github.com/projectcalico/vpp-dataplane) uses `extensions.projectcalico.org/selector` (e.g. [Calico-VPP Multinet services](https://github.com/projectcalico/vpp-dataplane/blob/v3.25.1/docs/multinet.md#multinet-services)). To simplify the user experience, a mutating webhook could read the selector, add them to the annotation and clear them from the specs when the type of service is detected.

Reviewer 4: (nit / optional for clarification)
```suggestion
Services without selectors will not get any EndpointSlice objects. Therefore, selecting pods can be done in different ways, for example, via annotation. An annotation will be used in the service to select which pods will be used as backend for this service. For example, [nokia/danm](https://github.com/nokia/danm) uses `danm.k8s.io/selector` (e.g. [DANM service declaration](https://github.com/nokia/danm/blob/v4.3.0/example/svcwatcher_demo/services/internal_lb_svc.yaml#L7)), and [projectcalico/vpp-dataplane](https://github.com/projectcalico/vpp-dataplane) uses `extensions.projectcalico.org/selector` (e.g. [Calico-VPP Multinet services](https://github.com/projectcalico/vpp-dataplane/blob/v3.25.1/docs/multinet.md#multinet-services)). To simplify the user experience, a mutating webhook could read the selector, add them to the annotation and clear them from the specs when the type of service is detected.

Please note that the mention in this KEP of `danm.k8s.io/selector` does not constitute an endorsement by the Kubernetes project of the unofficial `damn.k8s.io/selector` annotation key.
```

> Services without selectors will not get any EndpointSlice objects. Therefore, selecting pods can be done in different ways, for example, via annotation. An annotation will be used in the service to select which pods will be used as backend for this service. For example, [nokia/danm](https://github.com/nokia/danm) uses `danm.k8s.io/selector` (e.g. [DANM service declaration](https://github.com/nokia/danm/blob/v4.3.0/example/svcwatcher_demo/services/internal_lb_svc.yaml#L7)), and [projectcalico/vpp-dataplane](https://github.com/projectcalico/vpp-dataplane) uses `extensions.projectcalico.org/selector` (e.g. [Calico-VPP Multinet services](https://github.com/projectcalico/vpp-dataplane/blob/v3.25.1/docs/multinet.md#multinet-services)). To simplify the user experience, a mutating webhook could read the selector, add them to the annotation and clear them from the specs when the type of service is detected.

Reviewer 4: ```suggestion
Services without selectors will not get any EndpointSlice objects. Therefore, selecting pods can be done in different ways, for example, via annotation. An annotation will be used in the service to select which pods will be used as backend for this service. For example, [nokia/danm](https://github.com/nokia/danm) uses `danm.io/selector`, and [projectcalico/vpp-dataplane](https://github.com/projectcalico/vpp-dataplane) uses `extensions.projectcalico.org/selector` (e.g. [Calico-VPP Multinet services](https://github.com/projectcalico/vpp-dataplane/blob/v3.25.1/docs/multinet.md#multinet-services)). To simplify the user experience, a mutating webhook could read the selector, add them to the annotation and clear them from the specs when the type of service is detected.
```

> Services without selectors will not get any EndpointSlice objects. Therefore, selecting pods can be done in different ways, for example, via annotation. An annotation will be used in the service to select which pods will be used as backend for this service. For example, [nokia/danm](https://github.com/nokia/danm) uses `danm.k8s.io/selector` (e.g. [DANM service declaration](https://github.com/nokia/danm/blob/v4.3.0/example/svcwatcher_demo/services/internal_lb_svc.yaml#L7)), and [projectcalico/vpp-dataplane](https://github.com/projectcalico/vpp-dataplane) uses `extensions.projectcalico.org/selector` (e.g. [Calico-VPP Multinet services](https://github.com/projectcalico/vpp-dataplane/blob/v3.25.1/docs/multinet.md#multinet-services)). To simplify the user experience, a mutating webhook could read the selector, add them to the annotation and clear them from the specs when the type of service is detected.

Reviewer 2: This project last commit is from Sep 2022 https://github.com/nokia/danm/commits/master/

Reviewer 4: I think for a ruled-out alternative, referencing something stale is OK. The ecosystem doesn't have to buy in to the thing that we, uh, plan to reject.

Reviewer 2: I'm trying to gauge the importance of this feature and the demand, it seems there are no OSS implementations of this that are "alive" so I wonder if there are users ... or if there are only private implementations

The custom EndpointSlice Controller will then read the annotation to select the pods targeted by the service.

```yaml
apiVersion: v1
kind: Service
metadata:
  name: service
  annotations:

>     selector: "app=a"

Reviewer 4: How about:
```suggestion
    example.com/custom-service-selector: "app=a"
```

That way, it's not cluster-private.

spec: {}
```

This alternative potentially leads to confusion among users and inconsistency in how services are managed as each implementation is using its own annotation (see the nokia/danm and projectcalico/vpp-dataplane examples), leading to a fragmented approach.

> ### Well-known label: Use Dummy Selector

Reviewer 4: Try this:
```suggestion
### Private (per cluster) dummy selector label, custom controller
```
If you don't mean for the dummy selector to use a cluster-private label, you should revise the explanation.

Author: I am not sure what you mean here. the label in the selector are always within the cluster no?

The set of Pods targeted by a Service is determined by a selector, the labels in the selector must be included as part of the pod labels. If a dummy selector is added to the service, Kubernetes will not select any pod, the endpointslices created by Kubernetes will then be empty. To simplify the user experience, a mutating webhook could add the dummy selector when the type of service is detected.

The custom EndpointSlice Controller could read the service.spec.selector and ignore the dummy label to select pods targeted by the service.

```yaml
apiVersion: v1
kind: Service
metadata:
  name: service
spec:

>   selector:

Reviewer 4: This isn't a [well-known label](https://kubernetes.io/docs/reference/labels-annotations-taints/).

Author: If you refer to the title, sorry for the confusion, originally, the KEP had 2 proposals (well known label + generic endpointslice reconciler), Well-known Label indicates it is an alternative for the well-known label proposal.

Author: I removed it from the title of the alternative.

Reviewer 4: Got it.

  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

> This proposal adds a new well-known label `service.kubernetes.io/endpoint-controller-name` to Kubernetes Services. This label disables the default Kubernetes EndpointSlice controller for the services where this label is applied and delegates the control of EndpointSlices to a custom EndpointSlice controller.

Reviewer 2: we have the label `managed-by` already on the endpointslice controller and the ipaddress, I think we should be consistent and use the same here

```
rep -r managed-by staging/src/k8s.io/api
staging/src/k8s.io/api/discovery/v1/well_known_labels.go:       LabelManagedBy = "endpointslice.kubernetes.io/managed-by"
staging/src/k8s.io/api/discovery/v1beta1/well_known_labels.go:  LabelManagedBy = "endpointslice.kubernetes.io/managed-by"
staging/src/k8s.io/api/networking/v1alpha1/well_known_labels.go:        LabelManagedBy = "ipaddress.kubernetes.io/managed-by"
staging/src/k8s.io/api/networking/v1beta1/well_known_labels.go: LabelManagedBy = "ipaddress.kubernetes.io/managed-by"
```

Author: Any name is fine to me. For reference, the `service.kubernetes.io/service-proxy-name` is not added to the service. As highlighted in https://github.com/kubernetes/enhancements/pull/4771#discussion_r1685505890, a default controller name would have to be handled for the kubernetes controller manager.

Reviewer 2: yes, I added this comment just for opening the debate, I'd like to know if there are more thoughts on this

Reviewer 3: I think `endpointslice.kubernetes.io/managed-by` is probably fitting here. I'm not entirely sure whether we want this to apply to Endpoints as well though, if so, that could change what's best here.

Reviewer 4: This KEP really [does](https://github.com/kubernetes/enhancements/pull/4771#pullrequestreview-2379546600) need to explain how it interacts with the existing label.

Author: I am not sure what exactly has to be clarified with existing labels. This KEP proposes a label on the Service object, `endpointslice.kubernetes.io/managed-by` is already explained in the Kubernetes documentation.

Author: Thank you for the explanation, I tried to clarify the differences and how the `endpointslice.kubernetes.io/managed-by` label is impacted.

> Additionally, this KEP aims to give more flexibility for the EndpointSlice Reconciler module to allow users to reconcile EndpointSlices with any type of resources (Currently only Service/Pods is supported). For example, Endpoints could be supported for the Kubernetes EndpointSlice Mirroring Controller.

Reviewer 2: This part I can not really understand it, maybe is explained later ...

Reviewer 3: +1, "Endpoints could be supported for the Kubernetes EndpointSlice Mirroring Controller" is particularly confusing to me.

Delegating EndpointSlice control would allow custom controllers to define their own criterias for pod availability, selecting different pod IPs than the pod.status.PodIPs and more. As reference implementation, and since [KEP-3685](https://github.com/kubernetes/enhancements/issues/3685), the reconciler logic used by Kubernetes can be reused by custom EndpointSlice controllers.

Providing a generic EndpointSlice Reconciler module would allow users to reuse the same code as the Kubernetes EndpointSlice Controller, thus ensuring consistency and reducing the effort needed to implement custom reconciliation logic. As highlighted in this [comment](https://github.com/kubernetes/kubernetes/pull/118953#discussion_r1245970845), the EndpointSlice Mirroring Controller would also benefit from these changes since the generic EndpointSlice Reconciler could be used from the EndpointSlice Mirroring Controller.

### Goals

* Provide the ability to disable the Kubernetes EndpointSlice controller for particular services.
* Extend and enhance the EndpointSlice Reconciler module to make it generic.

> ### Non-Goals

Reviewer 2: a non goal is to add new features to the `EndpointSlice Mirroring Controller` , I can see there is some risk that sharing code, this shared code get new functionalities that we leak into the mirroring controller ... how are we planning to prevent that?

Author: I added the non-goal you mentioned. That a good point, I added `To avoid protential new functionalities to leak into other controllers re-using the EndpointSlice Reconciler (for example, the EndpointSlice Mirroring Controller), new functionnalities must be optional if possible.` to the `Risks and Mitigations` section, do you think this would be enough? Do you think this will be possible with any new potential features?

Reviewer 2: this is an important point that we need to be very clear, as it happened with other libraries , specially in client-go that the scope start to grow and makes the code unsustainable and introduce bugs on stable codebases.

 These controllers are built for kubernetes and to solve kubernetes problems, if the logic can be reused as a library perfect, Im +1 ... but adding options or flows that will be used only externally is a -1 for me.

If the multinetwork community wants to build new libraries on top, then we better add a new repo kubernetes-sigs/generic-endpointslice-controller and we add that code there

* Change / Replace / Deprecate the existing behavior of the Kubernetes EndpointSlice controller.
* Introduce additional supported types of the EndpointSlice controllers/Reconciler as part of Kubernetes.
* Modify the Service / EndpointSlice Specs.

## Proposal

#### Well-Known Label

`service.kubernetes.io/endpoint-controller-name` will be added as a well-known label applying on the Service object. When set on a service, the Endpoint, EndpointSlice, and EndpointSlice Mirroring for that service will be disabled, thus EndpointSlices for this service will not be created by the Kubernetes Controller Manager. 

#### Flexibility on the EndpointSlice Reconciler Module

> The reconciler structure will support more features to cover the requirements set by the current behavior of the EndpointSlice Mirroring Controller (e.g.: placeholder concept does not exist in the Mirroring Controller).

Reviewer 2: what is this placeholder concept?

Author: I added a paragraph in the section `Flexibility on the EndpointSlice Reconciler Module`.

The `Reconcile` function definition will be changed to accept general types of data (not specific to Services) and the list of features the endpointslices being reconciled will support (e.g. Traffic Distribution). Functions for specific types (Service/Pods for the EndpointSlice Controller and Endpoints for the EndpointSlice Mirroring Controller) generating data to pass to the `Reconcile` function will be provided by the EndpointSlice Reconciler module.

The Metrics and its cache between the EndpointSlice Reconciler and EndpointSlice Mirroring Reconciler are similar but not identical. They will be merged together to provide a re-usable and non-type specific metrics package.

The EndpointSlice controller will be adapted accordingly to support the new way of using the EndpointSlice Reconciler.

Finally, the EndpointSlice Mirroring reconciler will be discarded from `pkg/controller/endpointslicemirroring` and the EndpointSlice Mirroring Controller will use the EndpointSlice reconciler from staging with the appropriate configuration.

### User Stories (Optional)

#### Story 1

> As a Cloud Native Network Function (CNF) vendor, some of my services are handled by custom Service-Proxies/Gateways over secondary networks provided by, for example, [Multus](https://github.com/k8snetworkplumbingwg/multus-cni). IPs configured in the service and registered by the EndpointSlice controller must be only the secondary IPs provided by the secondary network provider (e.g. [Multus](https://github.com/k8snetworkplumbingwg/multus-cni), also see [KEP-3698](https://github.com/kubernetes/enhancements/issues/3698)).

Reviewer 2: ```suggestion
As a Cloud Native Network Function (CNF) vendor, some of my services are handled by custom Service-Proxies/Gateways over secondary networks provided by, for example, [Multus](https://github.com/k8snetworkplumbingwg/multus-cni). IPs configured in the service and registered by the EndpointSlice controller must be only the secondary IPs provided by the secondary network provider (e.g. [Multus](https://github.com/k8snetworkplumbingwg/multus-cni).
```

KEP-3698 is not approved and has still a lot of pending comments without answer, so using it as foundation of this KEP is not ideal

Author: I removed it.

> Therefore, it must be possible to disable the default Kubernetes EndpointSlice Controller for certain services and re-use the Kubernetes EndpointSlice reconciler implementation to create a controller for secondary network providers.

Reviewer 2: I assume that across this document, you mean EndpointSlice controller and Endpoint controller, or it will be an inconsistent behavior

Author: I fixed it, thank you.

### Notes/Constraints/Caveats (Optional)

N/A

### Risks and Mitigations

> The existing behavior will be kept by default, and the Kubernetes EndpointSlice Controller will be disabled only when the service contains this new label. This ensures services without the label to continue to be managed as usual.

Reviewer 2: nit, disabled sounds like the endpoint slice controller will be stopped, when we are really meaning the endpoint slice controller in the kube-controller-manager will omit this Service, maybe something as `the Kubernetes EndpointSlice Controller will not manage Services with the label ...`

Author: I fixed it, thank you.

This will have no effect on other EndpointSlice controller implementations since they will not be influenced by the presence of this label.

## Design Details

### Well-Known Label

> The kube-controller-manager should pass to the Endpoints, EndpointSlice and EndpointSlice Mirroring Controllers an informer selecting services that are not labelled with `service.kubernetes.io/endpoint-controller-name`.

Reviewer 2: See https://github.com/kubernetes/enhancements/pull/4771/files#r1685502488 , I think there is 2 things to flesh out here:
1. label name
2. default behavior, do we add a label with the default controller name or we assume no-label means default controller

Author: For reference, kube-proxy has no default controller name, non-labelled services are managed by kube-proxy.

### Flexibility on the EndpointSlice Reconciler Module

> New fields will be added to the `Reconciler` struct to cover the needs of the EndpointSlice and EndpointSlice Mirroring Controllers. Unlike the EndpointSlice Mirroring Controller, the EndpointSlice Controller needs placeholders if a service does not contain any endpoint. So a boolean `placeholderEnabled` will be added. The EndpointSlice Mirroring Controller does not verify which resource owns the EndpointSlice before Updating/Deleting it, another boolean `ownershipEnforced` would control this.

Reviewer 2: placeholder seems to require a section to explain better the concept https://github.com/kubernetes/enhancements/pull/4771/files#r1685505013

> The EndpointSlice Mirroring Controller does not verify which resource owns the EndpointSlice before Updating/Deleting it, another boolean `ownershipEnforced` would control this.

how do you deal with orphan objects then?

Author: The owner reference is related to this function in the reconciler: [staging/src/k8s.io/endpointslice/utils.go#L159](https://github.com/kubernetes/endpointslice/blob/v0.30.3/utils.go#L159). I am not sure why this is required, the EndpointSlice Controller doesn't use this to get the EndpointSlice(s) for a particalar service. It uses only the EndpointSlice labels [pkg/controller/endpointslice/endpointslice_controller.go#L369](https://github.com/kubernetes/kubernetes/blob/v1.30.3/pkg/controller/endpointslice/endpointslice_controller.go#L369). With the ownership required enabled, if an EndpointSlice does not have the owner set, then that EndpointSlice will be deleted (current behavior in the EndpointSlice Reconciler).

Reviewer 2: why should not be always the case? why should it be kept lingering if it is generated from a Service?

Author: I agree, this should always be the case. I am not sure why, but the EndpointSlice Mirroring Controller does not require this. The ownership is not enforced, if the EndpointSlice is not owned, the Reconciler will not delete it and create a correct owned one. I made the implementation, and the EndpointSlice Mirroring unit tests are not passing if we enforce the ownership.

Reviewer 2: I see, so is an implementation detail , can you send a PR and set the title to WIP so I can take a look?

Author: https://github.com/kubernetes/kubernetes/pull/126304

Sorry if the PR is so large, I was planning to create smaller PRs to propose these changes if this KEP would be accepted.

I moved the unit tests from here: [pkg/controller/endpointslicemirroring/reconciler_test.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/endpointslicemirroring/reconciler_test.go) to here: [staging/src/k8s.io/endpointslice/reconciler_test.go#L2274](https://github.com/LionelJouin/kubernetes/blob/endpointslice-generic/staging/src/k8s.io/endpointslice/reconciler_test.go#L2274)

If you add the option `WithOwnershipEnforced(true)` for the Reconciler used in the tests ([instanciated here](https://github.com/LionelJouin/kubernetes/blob/endpointslice-generic/staging/src/k8s.io/endpointslice/reconciler_test.go#L3484)), then the tests will fail.

To run the tests: `go test -count=1 -run ^TestReconcileWithEndpoints$ k8s.io/endpointslice`

Reviewer 2: @robscott can you please chime in? I'm worried that we do changes without fully understand the implications that bit us later ^^^

Reviewer 3: I get that once a bug has existed for long enough it can be a feature, but it's hard for me to justify any of our controllers ever deleting EndpointSlices they did not claim ownership of. I think a smallish bug fix PR for the EndpointSlice Mirroring controller just to address that would help simplify this PR.

Author: I mark this conversation as resolved as the flexibility of the reconciler as been removed from this KEP

    time.Time{}, // So endpoints.kubernetes.io/last-change-trigger-time will be removed.
)
```

#### Metrics

Most of the metrics exposed by the EndpointSlice and EndpointSlice Mirroring Controllers are similar. The EndpointSlice Controller exposes `endpointslices_changed_per_sync`, `syncs` and `services_count_by_traffic_distribution` that are not exposed by the EndpointSlice Mirroring Controller. The EndpointSlice Mirroring Controller exposes `endpoints_updated_per_sync`, `addresses_skipped_per_sync` and `endpoints_sync_duration` that are not exposed by the EndpointSlice Controller. Metrics exposed will need to be merged together, and the metrics package should be changed to support any type of resources (pods/service, endpoints...).

### Test Plan

[x] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

##### Unit tests

* Tests in `pkg/controller/endpointslicemirroring/reconciler_test.go` will be moved to `staging/src/k8s.io/endpointslice`

##### Integration tests

> TDB

Reviewer 2: I expect to see a good battery of these tests, are the more important for this feature

Author: This should be fixed now. E2E tests has been moved as integration and another one as been added to tests the feature gate when disabled.

##### e2e tests

> - Usage of `service.kubernetes.io/endpoint-controller-name` on services

Reviewer 2: these tests are better done in integration, there is no e2e functionality here , is only API operations

Author: Similar test for the `service.kubernetes.io/service-proxy-name` is an e2e test (https://github.com/kubernetes/kubernetes/blob/v1.30.3/test/e2e/network/service.go#L2246). Do you consider it different? Does this really has to be moved to integration test?

Reviewer 2: that test linked exercises the dataplane, if you were expecting to test also the dataplane then I stand corrected, but I'd like to see the copy of the api only tests in the integration too :)

Author: I understand now, thank you for the explanation, I moved it.

>     * A service is created with the label and the service has then no endpoints.

Reviewer 2: ```suggestion
    * A service is created with the label and the service has then no endpoints. Then service is updated removing the label and the service has endpoints.
```

>     * A service is created without the label, the service has endpoints. Then service is updated with the label and the service has no longer any endpoints.

Reviewer 2: ```suggestion
    * A service is created without the label, the service has endpoints and endpointslices. Then service is updated with the label and the service has no longer any endpoints or endpointslices.
```

Author: I fixed it, thank you.

### User Stories (Optional)

#### Story 1

As a Cloud Native Network Function (CNF) vendor, some of my services are handled by custom Service-Proxies/Gateways over secondary networks provided by, for example, [Multus](https://github.com/k8snetworkplumbingwg/multus-cni). IPs configured in the service and registered by the EndpointSlice controller must be only the secondary IPs provided by the secondary network provider (e.g. [Multus](https://github.com/k8snetworkplumbingwg/multus-cni)).

Therefore, it must be possible to disable the default Kubernetes Endpoints and EndpointSlice Controller for certain services and re-use the Kubernetes EndpointSlice reconciler implementation to create a controller for secondary network providers.

### Notes/Constraints/Caveats (Optional)

N/A

### Risks and Mitigations

The existing behavior will be kept by default, and the Kubernetes EndpointSlice Controller will not manage the Services with the label. This ensures services without the label to continue to be managed as usual.

This will have no effect on other EndpointSlice controller implementations since they will not be influenced by the presence of this label.

> To avoid protential new functionalities to leak into other controllers re-using the EndpointSlice Reconciler (for example, the EndpointSlice Mirroring Controller), new functionnalities must be optional if possible.

Reviewer 2: ```suggestion
To avoid potential new functionalities to leak into other controllers re-using the EndpointSlice Reconciler (for example, the EndpointSlice Mirroring Controller), new Kubernetes only functionalities must be optional if possible. No functionalities outside of kubernetes/kubernetes will be added to this controller.
```

Author: I fixed it, thank you.

  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

> 

Reviewer 6: In trying to address Antonio's comment: 

>  this KEP is missing some more explanation on the high level details

Maybe we can try to build out something here that's more like an executive summary that builds

I see this as two part, but, two related parts... IMO, summed up well by the title: "EndpointSlice controller flexibility"

1. We're looking to make the endpointslice reconciler more generic so that code can be reused by the Kubernetes EndpointSlice Mirroring Controller. 
2. Additionally, we're adding the well-known label, which gives further flexibility for custom controllers.

On use-cases, I definitely see this as a potentially important building block for services for multinetworking.

Here's my take on it, definitely input welcome, just trying to get the ball rolling:

```
This proposal aims to add flexibility for the EndpointSlice controller. These changes are two-fold, with two related changes: Making the EndpointSlices reconciler generic in such a way as to be reusable both by the EndpointSlices mirroring controller and externally, and also by adding a well-known label to allow for the possibility of external management of services by a custom controller. This well-known label functionality is a pattern that is present in components {insert here: endpoint slices, ipaddresses, etc}.  This functionality will prove to be useful for scenarios for multinetworking in Kubernetes where this functionality could be customized in external controllers, which may have a different set of rules for how services are handled when multiple networks are present.
```

Reviewer 2: We need more detail on some parts, so the proposal is to add a new label, but it does not explain how it interact with the existing APIs and semantics and components, per example:
1. can user add/remove the `service.kubernetes.io/endpoint-controller-name`  label to an existing Service? what are the code implications of handling updates? how the different endpoint slice controllers will reconcile? or do we plane to use the label only on Service creationg?
2. does this label applies to all type of Services or ClusterIP services only? NodePorts, LoadBalancer, Headless and ExternalName has complex semantics that has to be clearly defined
3. if apply to Loadbalancer how are the existing controllers supposed to handle the label? what happens with the DNS controllers? do they keep creating records from the new endpoints that most probably will not be reachable? how kube-proxy and Services proxies will be impacted, does it needs to omit these services?
5.  how is overlapping between endpoints reconcilers and IPs going to be handled?
7. what happens with Ingress and Gateway that consumes the Services object?

Reviewer 2: to clarify, what is missing is the technical details, not the goal, the goal is clear, but how to achieve that goal and what will be the implications is something we need to flesh out and describe in the KEP, is the most important part of the KEP process

I talked with @robscott yesterday, I just want to throw some thoughts here:

1. It has to be feature gated, to control upgrade and downgrades and avoid inconsistencies
2. Allow only to set the ownership label on Service creation, to simplify the logic, dealing with mutations on existing Services API is painful and has a lot of edge cases
3. It has to work on both Endpoints and EndpointSlices
4. We need to understand the possible security issues and implement guardrails https://github.com/kubernetes/kubernetes/issues/103675

Author: Thank you for the contribution and comments @dougbtv and @aojea

With this comment: https://github.com/kubernetes/enhancements/pull/4771#discussion_r1685503884, does the Summary @dougbtv proposed makes more sense?

> We need more detail on some parts, so the proposal is to add a new label, but it does not explain how it interact with the existing APIs and semantics and components, per example:
> 
> 1. can user add/remove the `service.kubernetes.io/endpoint-controller-name`  label to an existing Service? what are the code implications of handling updates? how the different endpoint slice controllers will reconcile? or do we plane to use the label only on Service creationg?
> 2. does this label applies to all type of Services or ClusterIP services only? NodePorts, LoadBalancer, Headless and ExternalName has complex semantics that has to be clearly defined
> 3. if apply to Loadbalancer how are the existing controllers supposed to handle the label? what happens with the DNS controllers? do they keep creating records from the new endpoints that most probably will not be reachable? how kube-proxy and Services proxies will be impacted, does it needs to omit these services?
> 4. how is overlapping between endpoints reconcilers and IPs going to be handled?
> 5. what happens with Ingress and Gateway that consumes the Services object?

1. I clarified these points in the latest update.
2. I clarified these points in the latest update.
3. I think this is out of scope for this KEP since it is already possible to create Endpoints/EndpointSlices from another controller, so this problem is already existing. Kube-proxy and DNS controllers will register "Unknown" IPs. Maybe Kube-proxy and DNS controllers should be documented to be clear which IPs are supported? (For reference KNPG could work with Multus, I made a PoC about it here: https://github.com/LionelJouin/l-3-4-gateway-api-poc)
4. I am not sure if I understood this point. Could you give more details?
5. Similar to 3., it's already possible to create Endpoints and EndpointSlices with your custom IPs. Ingress and Gateway depend on specific implementation, so maybe these implementations must be documented what kind of IPs are supported?

I am not sure if and what I am missing anything on the implications. I will try to read more about the NodePorts, LoadBalancer, Headless and ExternalName services you mentioned and see if anything must be added.

> to clarify, what is missing is the technical details, not the goal, the goal is clear, but how to achieve that goal and what will be the implications is something we need to flesh out and describe in the KEP, is the most important part of the KEP process
> 
> I talked with @robscott yesterday, I just want to throw some thoughts here:
> 
> 1. It has to be feature gated, to control upgrade and downgrades and avoid inconsistencies
> 2. Allow only to set the ownership label on Service creation, to simplify the logic, dealing with mutations on existing Services API is painful and has a lot of edge cases
> 3. It has to work on both Endpoints and EndpointSlices
> 4. We need to understand the possible security issues and implement guardrails [CVE-2021-25740: Endpoint & EndpointSlice permissions allow cross-Namespace forwarding kubernetes#103675](https://github.com/kubernetes/kubernetes/issues/103675)

1. What would need to be feature gated? 
    * The EndpointSlice Controller should not receive any feature except some metrics changes (To align with the EndpointSlice Mirroring Controller). 
    * The EndpointSlice Mirroring controller will use the common reconciler. This could be feature-gated, the old reconciler would be used without the feature-gate enabled. Otherwise the controller should not receive any feature except some metrics changes (To align with the EndpointSlice Controller).
    * The label could eventually be feature-gated too.
2. The idea was to allow mutation of the label. The Endpoints/EndpointSlice should only be added/removed when adding/deleting the label on the service.
3. Yes, this is planned.
4. I still need to read this more in detail to understand it.

Reviewer 2: > 1. What would need to be feature gated?

why should we not? this is adding a new behavior and I exposed some of things we already do not know and we need feedback, this will also allow us to have a escape hatch if after a few releases we found out there are problems with this approach, as possible security exploits ... there are a lot of unknowns ... I feel pretty strong about requiring a feature gate for adding this behavior

> 2.  The idea was to allow mutation of the label.

what is the use case for this? why someone will set a service and then mutate it to use it by another controller? what happens with the old Endpoints and Endpointslices, do we leave them orphan?
We already have an unsolved issue because of this https://github.com/kubernetes/kubernetes/issues/118376 ,

Delegating EndpointSlice control would allow custom controllers to define their own criteria for pod availability, selecting different pod IPs than the pod.status.PodIPs and more. As a reference implementation, and since [KEP-3685](https://github.com/kubernetes/enhancements/issues/3685), the reconciler logic used by Kubernetes can be reused by custom EndpointSlice controllers.

Providing a generic EndpointSlice Reconciler module would allow users to reuse the same code as the Kubernetes EndpointSlice Controller, thus ensuring consistency and reducing the effort needed to implement custom reconciliation logic. As highlighted in this [comment](https://github.com/kubernetes/kubernetes/pull/118953#discussion_r1245970845), the EndpointSlice Mirroring Controller would also benefit from these changes since the generic EndpointSlice Reconciler could be used from the EndpointSlice Mirroring Controller.

### Goals

> * Provide the ability to disable the Kubernetes EndpointSlice controller for particular services.

Reviewer 3: Reading further below, it sounds like you want to disable both Endpoints and EndpointSlice controllers, recommend stating that here.

Author: I fixed it, thank you.

> * Extend and enhance the EndpointSlice Reconciler module to make it generic.

Reviewer 3: This feels more like a nice implementation detail than a goal, what are we trying to accomplish? I think this is tied to making the code more reusable to make custom EndpointSlice controllers easier to maintain, but it would be good to clarify.

### Non-Goals

* Change / Replace / Deprecate the existing behavior of the Kubernetes EndpointSlice controller.
* Introduce additional supported types of the EndpointSlice controllers/Reconciler as part of Kubernetes.
* Modify the Service / EndpointSlice Specs.
* Add new features to the EndpointSlice Mirroring Controller.

## Proposal

#### Well-Known Label

`service.kubernetes.io/endpoint-controller-name` will be added as a well-known label applying on the Service object. When set on a service, no matter the service specs, the Endpoint, EndpointSlice, and EndpointSlice Mirroring controllers for that service will be disabled, thus Endpoints and EndpointSlices for this service will not be created by the Kubernetes Controller Manager. If the label is not set, the Endpoint, EndpointSlice, and EndpointSlice Mirroring controllers will be enabled for that service and the Endpoints and EndpointSlices will be handled as of today.

> The Kubernetes Controller Manager will implement this label both at object creation and on dynamic addition/removal/updates of this label.

Reviewer 3: This seems a bit confusing. I generally think of KCM as the broader infrastructure in which our controllers run, but each EndpointSlice controller would ultimately be responsible for setting this label right?

Author: No, the user has to set the label, then KCM has to reconcile the EndpointSlice only for the service without the label. The user, can if he wants, update the Label of the service, then KCM will have to react on that update.

Reviewer: ```suggestion
The endpoint-related controllers will obey this label both at object creation and on dynamic addition/removal/updates of this label.
```

#### Flexibility on the EndpointSlice Reconciler Module

The reconciler structure will support more features to cover the requirements set by the current behavior of the EndpointSlice Mirroring Controller (e.g.: placeholder EndpointSlice does not exist in the Mirroring Controller).

> The `Reconcile` function definition will be changed to accept general types of data (not specific to Services) and the list of features the endpointslices being reconciled will support (e.g. Traffic Distribution). Functions for specific types (Service/Pods for the EndpointSlice Controller and Endpoints for the EndpointSlice Mirroring Controller) generating data to pass to the `Reconcile` function will be provided by the EndpointSlice Reconciler module.

Reviewer 3: This sounds particularly risky, we'll need to tread very carefully here when changing the signature of such a fundamental function. Worse, much of our test coverage is built around this function, so it would be hard to keep the same tests through this change.

Author: The unit tests are not complex to modify, a helper function is used (https://github.com/kubernetes/endpointslice/blob/v0.30.3/reconciler_test.go#L2313). I changed it in the first version I made: https://github.com/LionelJouin/kubernetes/blob/endpointslice-generic/staging/src/k8s.io/endpointslice/reconciler_test.go#L3437. This is for Pod/Service objects.

To reconcile Endpoints, we could just use another helper function: https://github.com/LionelJouin/kubernetes/blob/endpointslice-generic/staging/src/k8s.io/endpointslice/reconciler_test.go#L3677.

> The Metrics and its cache between the EndpointSlice Reconciler and EndpointSlice Mirroring Reconciler are similar but not identical. They will be merged together to provide a re-usable and non-type specific metrics package.

Reviewer 3: I'm not really sure what kind of backwards compatibility guarantees we offer for metrics, but this also sounds quite risky.

Author: "Alpha metrics do not have any API guarantees. These metrics must be used at your own risk, subsequent versions of Kubernetes may remove these metrics altogether, or mutate the API in such a way that breaks existing dashboards and alerts." https://kubernetes.io/docs/reference/instrumentation/metrics/#list-of-alpha-kubernetes-metrics

### User Stories (Optional)

#### Story 1

As a Cloud Native Network Function (CNF) vendor, some of my services are handled by custom Service-Proxies/Gateways over secondary networks provided by, for example, [Multus](https://github.com/k8snetworkplumbingwg/multus-cni). IPs configured in the service and registered by the EndpointSlice controller must be only the secondary IPs provided by the secondary network provider (e.g. [Multus](https://github.com/k8snetworkplumbingwg/multus-cni)).

Therefore, it must be possible to disable the default Kubernetes Endpoints and EndpointSlice Controller for certain services and re-use the Kubernetes EndpointSlice reconciler implementation to create a controller for secondary network providers.

### Notes/Constraints/Caveats (Optional)

N/A

### Risks and Mitigations

The existing behavior will be kept by default, and the Kubernetes EndpointSlice Controller will not manage the Services with the label. This ensures services without the label to continue to be managed as usual.

> This will have no effect on other EndpointSlice controller implementations since they will not be influenced by the presence of this label.

Reviewer 3: I think the biggest risk is that we could introduce significant bugs because this is making significant changes to the code behind a critical GA controller.

Reviewer 2: There is also another risk, controllers that rely on Services that consume the Pods directly instead of the EndpointSlices (cc @bowei @swetharepakula as the ingress-gce does this for NEG) will need to be modified for these services.

Reviewer 5: This counts as a point in favor of "use an empty selector".  But also, that just seems broken. :)

### Test Plan

[x] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

##### Unit tests

* Tests in `pkg/controller/endpointslicemirroring/reconciler_test.go` will be moved to `staging/src/k8s.io/endpointslice`

##### Integration tests

- Usage of `service.kubernetes.io/endpoint-controller-name` on services
    * A service is created with the label and the service has then no endpoints neither endpointslices. Then service is updated removing the label and the service has now endpoints and endpointslices.
    * A service is created without the label, the service has endpoints and endpointslices. Then service is updated with the label and the service has no longer any endpoints nor endpointslices.

##### e2e tests

> TDB

Reviewer 2: Sorry, my previous comment was confusing, we need at least one or two e2e test that implementation use to validate the expected behavior, implementations must be consistent

Reviewer 2: this is still missing, 

"A service is created with the label and an external controller creates EndpointSlices for that Service, Pods connecting to that service should be able to connect to the custom crafted endpoint"

For this we just can create the fake endpointslices manually on the e2e test emulating the behavior of an external controller and pointing them to some random pod, there are already some tests with this logic you can reuse

Author: Thank you for the suggestion, I added it.

###### Does this feature depend on any specific services running in the cluster?

No

### Scalability

###### Will enabling / using this feature result in any new API calls?

No

###### Will enabling / using this feature result in introducing new API types?

No

###### Will enabling / using this feature result in any new calls to the cloud provider?

No

###### Will enabling / using this feature result in increasing size or count of the existing API objects?

No

###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?

No

###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?

No

###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?

No

### Troubleshooting

###### How does this feature react if the API server and/or etcd is unavailable?

N/A

###### What are other known failure modes?

N/A

###### What steps should be taken if SLOs are not being met to determine the problem?

N/A

## Implementation History

- Initial proposal: 2024-07-19

## Drawbacks

TBD

> ## Alternatives

Reviewer 2: Another alternative is to disable the kube-controller-manager endpointslice controller completely and use only the external one

https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/

```
--controllers strings     Default: "*"
A list of controllers to enable. '*' enables all on-by-default controllers, 'foo' enables the controller named 'foo', '-foo' disables the controller named 'foo'.
All controllers: bootstrap-signer-controller, certificatesigningrequest-approving-controller, certificatesigningrequest-cleaner-controller, certificatesigningrequest-signing-controller, cloud-node-lifecycle-controller, clusterrole-aggregation-controller, cronjob-controller, daemonset-controller, deployment-controller, disruption-controller, endpoints-controller, endpointslice-controller, endpointslice-mirroring-controller, ephemeral-volume-controller, garbage-collector-controller, horizontal-pod-autoscaler-controller, job-controller, legacy-serviceaccount-token-cleaner-controller, namespace-controller, node-ipam-controller, node-lifecycle-controller, node-route-controller, persistentvolume-attach-detach-controller, persistentvolume-binder-controller, persistentvolume-expander-controller, persistentvolume-protection-controller, persistentvolumeclaim-protection-controller, pod-garbage-collector-controller, replicaset-controller, replicationcontroller-controller, resourceclaim-controller, resourcequota-controller, root-ca-certificate-publisher-controller, service-cidr-controller, service-lb-controller, serviceaccount-controller, serviceaccount-token-controller, statefulset-controller, storage-version-migrator-controller, storageversion-garbage-collector-controller, taint-eviction-controller, token-cleaner-controller, ttl-after-finished-controller, ttl-controller, validatingadmissionpolicy-status-controller
Disabled-by-default controllers: bootstrap-signer-controller, token-cleaner-controller
```

Author: Thank you, I added it to the alternatives.
